superdesk/audit/commands.py

```diff
@@ -0,0 +1,158 @@
+                                          {'extra.item': {'$exists': True}}
+                                          ]}]}
+
+    not_item_entry_query = {'$or': [{'resource': {'$nin': item_resources}}, {'extra': {'$exists': False}},
+                                    {'$and': [{'extra': {'$exists': True}},
+                                              {'extra.guid': {'$exists': False}},
+                                              {'extra._id': {'$exists': False}},
+                                              {'extra.item_id': {'$exists': False}},
+                                              {'extra.item': {'$exists': False}}
+                                              ]}]}
+
+    def run(self):
+        if config.AUDIT_EXPIRY_MINUTES == 0:
+            logger.info('Audit purge is not eanabled')
+            return
+        logger.info("Starting audit purge")
+        self.expiry = utcnow() - datetime.timedelta(minutes=config.AUDIT_EXPIRY_MINUTES)
+        self.purge_old_entries()
+        self.purge_orphaned_item_audits()
+        logger.info("Completed audit purge")
+
+    def _get_archive_ids(self, ids):
+        """
+        Given a set of ids return those that have an entry in the archive collection
+        :param ids:
+        :return:
+        """
+        service = superdesk.get_resource_service('archive')
+        query = {'_id': {'$in': list(ids)}}
+        req = ParsedRequest
+        req.projection = '{"_id": 1}'
+        archive_ids = service.get_from_mongo(req=req, lookup=query)
+        existing = list([item['_id'] for item in archive_ids])
+        return set(existing)
+
+    def _extract_item_id(self, item):
+        """
+        Given an audit item try to extract the id of the item that it relates to
+        :param item:
+        :return:
+        """
+        try:
+            if 'guid' in item['extra']:
+                return item['extra']['guid']
+            elif '_id' in item['extra']:
+                return item['extra']['_id']
+            elif 'item_id' in item['extra']:
+                return item['extra']['item_id']
+            elif 'item' in item['extra']:
+                return item['extra']['item']
+        except:
+            return None
+        return None
+
+    def purge_orphaned_item_audits(self):
+        """
+        Purge the audit items that do not have associated entries existing in archive
+        :return:
+        """
+        service = superdesk.get_resource_service('audit')
+        current_id = None
+
+        # Scan the audit collection for items to delete
+        while True:
+            query = self.item_entry_query
+            query['$and'].append({'_updated': {'$lte': date_to_str(self.expiry)}})
+            if current_id:
+                query['$and'].append({'_id': {'$gt': current_id}})
+            req = ParsedRequest
+            req.sort = '[("_id", 1)]'
+            req.projection = '{"_id": 1, "extra.guid": 1, "extra._id": 1, "extra.item_id": 1, "extra.item": 1}'
```

Comment:
It should be possible to run it as a command, I was thinking to set the expiry to a value that would delete a sane number and run the command and repeat. 

We'll need some indexes on the audit collection. I guess we can import the production audit collection into UAT and tune it there, the fact that there would be no related items in the archive shouldn't matter.